tests/webbrowser/test_webbrowser.py

```diff
@@ -0,0 +1,142 @@
+            nursery.cancel_scope.cancel()
+
+        assert process.poll() == (1 if is_win32 else -SIGTERM), "Process has been terminated"
+        assert [(record.name, record.levelname, record.msg) for record in caplog.records] == [
+            ("streamlink.webbrowser.webbrowser", "debug", "Waiting for web browser process to terminate"),
+        ]
+
+    @pytest.mark.trio()
+    async def test_terminate_on_nursery_timeout(self, caplog: pytest.LogCaptureFixture, mock_clock, webbrowser_launch):
+        nursery: trio.Nursery
+        process: trio.Process
+        async with webbrowser_launch(timeout=10) as (nursery, process):  # noqa: F841
+            assert process.poll() is None, "process is still running"
+            mock_clock.jump(20)
+            await trio.sleep(0)
+
+        assert process.poll() == (1 if is_win32 else -SIGTERM), "Process has been terminated"
+        assert [(record.name, record.levelname, record.msg) for record in caplog.records] == [
+            ("streamlink.webbrowser.webbrowser", "warning", "Web browser task group has timed out"),
+            ("streamlink.webbrowser.webbrowser", "debug", "Waiting for web browser process to terminate"),
+        ]
+
+    @pytest.mark.trio()
+    async def test_terminate_on_nursery_baseexception(self, caplog: pytest.LogCaptureFixture, webbrowser_launch):
+        class FakeBaseException(BaseException):
+            pass
+
+        nursery: trio.Nursery
+        process: trio.Process
+        with pytest.raises(FakeBaseException):  # noqa: PT012
+            async with webbrowser_launch() as (nursery, process):  # noqa: F841
+                assert process.poll() is None, "process is still running"
+                raise FakeBaseException()
+
+        assert process.poll() == (1 if is_win32 else -SIGTERM), "Process has been terminated"
+        assert [(record.name, record.levelname, record.msg) for record in caplog.records] == [
+            ("streamlink.webbrowser.webbrowser", "debug", "Waiting for web browser process to terminate"),
+        ]
+
+    @pytest.mark.trio()
+    # don't check for non-zero exit codes - we don't care
+    @pytest.mark.parametrize("exit_code", [0, 1])
+    async def test_process_ended_early(self, caplog: pytest.LogCaptureFixture, webbrowser_launch, exit_code):
+        nursery: trio.Nursery
+        process: trio.Process
+        async with webbrowser_launch(timeout=10) as (nursery, process):  # noqa: F841
+            assert process.poll() is None, "process is still running"
+            assert process.stdin
+            await process.stdin.send_all(str(exit_code).encode() + b"\r\n")
+            await trio.sleep(5)
+
+        assert process.poll() == exit_code, "Process has ended with the right exit code"
+        assert [(record.name, record.levelname, record.msg) for record in caplog.records] == [
+            ("streamlink.webbrowser.webbrowser", "warning", "Web browser process ended early"),
+        ]
```

Comment:
Apparently there's an issue with this test on Windows.

One CI job of the dependent PR #5388 which includes this branch has failed:
https://github.com/streamlink/streamlink/actions/runs/5335524587/jobs/9668866676?pr=5388#step:5:271

> ```
> >       assert process.poll() == exit_code, "Process has ended with the right exit code"
> E       AssertionError: Process has ended with the right exit code
> E       assert 1 == 0
> E        +  where 1 = <bound method Process.poll of <trio.Process ['C:\\hostedtoolcache\\windows\\Python\\3.7.9\\x64\\python.exe', '-c', 'import sys; sys.exit(int(sys.stdin.readline()))']: exited with status 1>>()
> E        +    where <bound method Process.poll of <trio.Process ['C:\\hostedtoolcache\\windows\\Python\\3.7.9\\x64\\python.exe', '-c', 'import sys; sys.exit(int(sys.stdin.readline()))']: exited with status 1>> = <trio.Process ['C:\\hostedtoolcache\\windows\\Python\\3.7.9\\x64\\python.exe', '-c', 'import sys; sys.exit(int(sys.stdin.readline()))']: exited with status 1>.poll
> ```

No idea what's going on here...

The fake web browser process is this (from the `webbrowser_launch` fixture in `conftest.py`)
```sh
python -c 'import sys; sys.exit(int(sys.stdin.readline()))'
```
and it's supposed to exit with code 0 when `0\r\n` is written on its stdin pipe within the 5 second window (`sys.stdin.readline()` blocks).

Why is the process exiting with code 1? This looks like a generic exit code on Windows. It's the same when terminating the process, hence why we can't check for `-SIGTERM` in the other tests. Did the pipe break or what's going on here?

I've been running the webbrowser tests on my Arch Linux system in a loop for a while, and there are no timing issues. Guess I will have to debug this race condition in my Windows VM. :vomiting_face: 

Help or hints are appreciated...

Reply:
I had the `tests/webbrowser` tests now running in a loop in my Windows VM for over 45 mins.... More than a thousand iterations without failures.
```fish
while pytest --disable-warnings -qq tests/webbrowser/; end
```
I'll ignore this failure for now and if it comes up again, I will ignore the test on Windows.